lib/spack/spack/solver/asp.py

```diff
@@ -525,12 +527,14 @@ def _compute_specs_from_answer_set(self):
                 node = SpecBuilder.make_node(pkg=providers[0])
             candidate = answer.get(node)
 
-            if candidate and candidate.build_spec.satisfies(input_spec):
-                if not candidate.satisfies(input_spec):
-                    tty.warn(
-                        "explicit splice configuration has caused the concretized spec"
-                        f" {candidate} not to satisfy the input spec {input_spec}"
-                    )
+            if candidate and candidate.satisfies(input_spec):
+                self._concrete_specs.append(answer[node])
+                self._concrete_specs_by_input[input_spec] = answer[node]
+            elif candidate and candidate.build_spec.satisfies(input_spec):
+                tty.warn(
+                    "explicit splice configuration has caused the concretized spec"
+                    f" {candidate} not to satisfy the input spec {input_spec}"
+                )
```

Comment:
Your intuition is correct. Splicing can produce a satisfying spec, where it's build spec did not satisfy.

Reply:
To answer your other question, the features are largely orthogonal, so can technically exist together in the context of the same solve.

Reply:
> To answer your other question, the features are largely orthogonal, so can technically exist together in the context of the same solve.

If used together, couldn't they possibly drive the result into the else branch by "chaining" automatic and explicit splicing (which would probably be hard to debug at that point)? 

The scenario I'm thinking about is the following:
- A user makes a request, say for `^intel-mpi`
- Automatic splicing has a rule that allows splicing `^mpich` for `^intel-mpi`. So a spec originally built with `mpich` gets spliced and now contains `intel-mpi`.
- Explicit splicing has a rule to splice `^intel-mpi` for, say, `^mvapich`

At that point, if I understand correctly, we would have both the concrete spec and its build spec not satisying the user request.[^1]

[^1]: I still have to check the implementation in details, but was wondering if that is the expected outcome of this situation.

Reply:
> If used together, couldn't they possibly drive the result into the else branch by "chaining" automatic and explicit splicing (which would probably be hard to debug at that point)? 

I never really thought about that case, and it certainly _could_ happen and would cause a crash; however, I have a few responses:
1. Explicit splicing is definitely and unsafe feature. One is basically asserting that they know better than the concretizer, and making arbitrary post-hoc changes to concretized specs. 
2. Splicing, conceptually not necessarily in this implementation, is transitive. There would be no reason to explicitly splice `mvapich` into a spliced `intel-mpi`. the user should just splice directly into `mpich`. Since they control the explicit splicing constraints, they can fix that if it is a problem. 

All of this is to say that I don't necessarily think there is a reason for the features to be mutually exclusive; explicit splicing can be used while package developers catch up to adding `can_splice` constraints, but automatic splicing should eventually eclipse all non-research use cases of explicit splicing once all necessary `can_splice` directives are added to spliceable packages. 

I think my next project, which is beyond the scope of this PR, will be automatically adding `can_splice` facts to packages based on `libabigail`/some bespoke ABI discovery tool.

Reply:
> There would be no reason to explicitly splice mvapich into a spliced intel-mpi. the user should just splice directly into mpich. Since they control the explicit splicing constraints, they can fix that if it is a problem.

Sure. What I'm saying is not that user would _willingly_ do a transitive splice (first automatic, followed by an explicit one), but that this situation could accidentally happen if they use both together. We should at least be aware of that. 

> I think my next project, which is beyond the scope of this PR, will be automatically adding can_splice facts to packages based on libabigail/some bespoke ABI discovery tool.

:+1:

Reply:
Ah, I see. Yeah, we could say "use these with disjoint virtuals if you want to use them together" but in general *heavily* recommend against doing that. 

I'm currently working on an extension that users can control which packages they want automatic splicing to be turned on for, but that is not yet ready, since the exact subsumption semantics are a little tricky.